dbms/src/Storages/StorageODBC.cpp

```diff
@@ -1,73 +1,161 @@
+    , remote_database_name(remote_database_name_)
+    , remote_table_name(remote_table_name_)
+{
+    const auto & config = context_.getConfigRef();
+    size_t bridge_port = config.getUInt("odbc_bridge.port", 9018);
+    std::string bridge_host = config.getString("odbc_bridge.host", "localhost");
+
+    uri.setHost(bridge_host);
+    uri.setPort(bridge_port);
+    uri.setScheme("http");
+
+    ping_uri = uri;
+    ping_uri.setPath("/ping");
+}
+
+std::string StorageODBC::getReadMethod() const
+{
+    return Poco::Net::HTTPRequest::HTTP_POST;
+}
+
+std::vector<std::pair<std::string, std::string>> StorageODBC::getReadURIParams(const Names & column_names,
+    const SelectQueryInfo & /*query_info*/,
+    const Context & /*context*/,
+    QueryProcessingStage::Enum & /*processed_stage*/,
+    size_t max_block_size) const
 {
-    pool = createAndCheckResizePocoSessionPool([&]
+    NamesAndTypesList cols;
+    for (const String & name : column_names)
     {
-        return std::make_shared<Poco::Data::SessionPool>("ODBC", validateODBCConnectionString(connection_string));
-    });
+        auto column_data = getColumn(name);
+        cols.emplace_back(column_data.name, column_data.type);
+    }
+    std::vector<std::pair<std::string, std::string>> result;
+
+    result.emplace_back("connection_string", connection_string);
+    result.emplace_back("columns", cols.toString());
+    result.emplace_back("max_block_size", std::to_string(max_block_size));
+
+    return result;
 }
 
-BlockInputStreams StorageODBC::read(
-    const Names & column_names,
+std::function<void(std::ostream &)> StorageODBC::getReadPOSTDataCallback(const Names & /*column_names*/,
     const SelectQueryInfo & query_info,
     const Context & context,
-    QueryProcessingStage::Enum & processed_stage,
-    size_t max_block_size,
-    unsigned /*num_streams*/)
+    QueryProcessingStage::Enum & /*processed_stage*/,
+    size_t /*max_block_size*/) const
 {
-    check(column_names);
-    processed_stage = QueryProcessingStage::FetchColumns;
     String query = transformQueryForExternalDatabase(
         *query_info.query, getColumns().ordinary, IdentifierQuotingStyle::DoubleQuotes, remote_database_name, remote_table_name, context);
 
-    Block sample_block;
-    for (const String & name : column_names)
+    return [query](std::ostream & os) { os << "query=" << query; };
+}
+
+bool StorageODBC::checkODBCBridgeIsRunning() const
+{
+    try
     {
-        auto column_data = getColumn(name);
-        sample_block.insert({ column_data.type, column_data.name });
+        ReadWriteBufferFromHTTP buf(ping_uri, Poco::Net::HTTPRequest::HTTP_GET, nullptr);
+        return checkString("Ok.", buf);
+    }
+    catch (...)
+    {
+        return false;
+    }
+}
+
+
+void StorageODBC::startODBCBridge() const
+{
+    const auto & config = context_global.getConfigRef();
+    const auto & settings = context_global.getSettingsRef();
+    Poco::Path path{config.getString("application.dir", "")};
```

Comment:
What is `application.dir`?

Reply:
https://github.com/ClickHouse-Extras/poco/blob/3df947389e6d9654919002797bdd86ed190b3963/Util/include/Poco/Util/Application.h#L76 -- It stores information about executable path. I don't know other reliable and cross-platform (linux and macOS) ways to get it from source code.

Reply:
Ok.